src/handlers/VehicleTestController.ts

```diff
@@ -1,58 +1,230 @@
+        vehicleWheels: payload.numberOfWheelsDriven,
+      };
+      const testTypesWithTestCodesAndClassification = await this.dataProvider.getTestTypesWithTestCodesAndClassification(
+        payload.testTypes,
+        testTypeParams
+      );
+      payload.testTypes = testTypesWithTestCodesAndClassification;
+      const payloadWithTestNumber = await this.dataProvider.setTestNumberForEachTestType(
+        payload
+      );
+      const payloadWithExpiryDate = await this.generateExpiryDate(
+        payloadWithTestNumber
+      );
+      const payloadWithCertificateNumber = VehicleTestController.AssignCertificateNumberToTestTypes(
+        payloadWithExpiryDate
+      );
+      const payloadWithAnniversaryDate = VehicleTestController.calculateAnniversaryDate(
+        payloadWithCertificateNumber
+      );
+      payloadWithAnniversaryDate.vehicleId = payloadWithAnniversaryDate.vrm;
+      const result = await this.dataProvider.insertTestResult(payloadWithAnniversaryDate);
+      return result;
+    } catch (error) {
+      if (error.statusCode === 400 && error.message === enums.MESSAGES.CONDITIONAL_REQUEST_FAILED) {
+          console.info(
+            "TestResultService.insertTestResult: Test Result id already exists",
+            error
+          );
+          error =  new models.HTTPResponse(201, enums.MESSAGES.ID_ALREADY_EXISTS);
+        }
+      throw error;
+    }
+  }
+//#endregion
+//#region [rgba(0, 205, 30, 0.1)] Private functions
+  /**
+   * Note: When performing actions on a moment instance, it gets mutated
+   * Note: Expiry dates on the payload should be set at the UTC start of day.
+   *
+   * @param payload
+   */
+  private async generateExpiryDate(
+    payload: models.ITestResultPayload
+  ): Promise<models.ITestResultPayload> {
+    try {
+      if (
+        payload.testStatus !== enums.TEST_STATUS.SUBMITTED ||
+        utils.ValidationUtil.isNotAllowedVehicleTypeForExpiry(
+          payload.vehicleType
+        )
+      ) {
+        return Promise.resolve(payload);
+      }
+      const expiryTestTypes = payload.testTypes.filter((testType) =>
+        utils.ValidationUtil.isAllowedTestTypeForExpiry(testType)
+      );
+
+      const recentExpiry = await this.dataProvider.getMostRecentExpiryDate(
+        payload.systemNumber
+      );
+
+      expiryTestTypes.forEach((testType: any, index: number) => {
+        const testTypeForExpiry: models.TestTypeForExpiry = {
+          testType,
+          vehicleType:
+            enums.VEHICLE_TYPE[
+              payload.vehicleType.toUpperCase() as keyof typeof enums.VEHICLE_TYPE
+            ],
+          recentExpiry,
+          regnOrFirstUseDate: VehicleTestController.getRegistrationOrFirstUseDate(payload),
+          hasHistory: !DateProvider.isSameAsEpoc(recentExpiry),
+          hasRegistration: DateProvider.isValidDate(
+            VehicleTestController.getRegistrationOrFirstUseDate(payload)
+          ),
+        };
+        console.log(testTypeForExpiry);
```

Comment:
Do we want to add another console.log to show what we are logging out?

Reply:
added log title